gitlab/resource_deploytoken.go

```diff
@@ -0,0 +1,168 @@
+limitations under the License.
+*/
+
+package gitlab
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/xanzy/go-gitlab"
+
+	"github.com/fluxcd/go-git-providers/gitprovider"
+	"github.com/fluxcd/go-git-providers/validation"
+)
+
+func newDeployToken(c *DeployTokenClient, token *gitlab.DeployToken) *deployToken {
+	return &deployToken{
+		k: *token,
+		c: c,
+	}
+}
+
+var _ gitprovider.DeployToken = &deployToken{}
+
+type deployToken struct {
+	k gitlab.DeployToken
+	c *DeployTokenClient
+}
+
+func (dk *deployToken) Get() gitprovider.DeployTokenInfo {
+	return deployTokenFromAPI(&dk.k)
+}
+
+func (dk *deployToken) Set(info gitprovider.DeployTokenInfo) error {
+	if err := info.ValidateInfo(); err != nil {
+		return err
+	}
+	deployTokenInfoToAPIObj(&info, &dk.k)
+	return nil
+}
+
+func (dk *deployToken) APIObject() interface{} {
+	return &dk.k
+}
+
+func (dk *deployToken) Repository() gitprovider.RepositoryRef {
+	return dk.c.ref
+}
+
+// Update will apply the desired state in this object to the server.
+// Only set fields will be respected (i.e. PATCH behaviour).
+// In order to apply changes to this object, use the .Set({Resource}Info) error
+// function, or cast .APIObject() to a pointer to the provider-specific type
+// and set custom fields there.
+//
+// ErrNotFound is returned if the resource does not exist.
+//
+// The internal API object will be overridden with the received server data.
+func (dk *deployToken) Update(ctx context.Context) error {
+	// Delete the old token and recreate
+	if err := dk.Delete(ctx); err != nil {
+		return err
+	}
+	return dk.createIntoSelf()
+}
+
+// Delete deletes a deploy token from the repository.
+//
+// ErrNotFound is returned if the resource does not exist.
+func (dk *deployToken) Delete(_ context.Context) error {
+	// We can use the same DeployToken ID that we got from the GET calls. Make sure it's non-nil.
+	// This _should never_ happen, but just check for it anyways to avoid panicing.
+	if dk.k.ID == 0 {
+		return fmt.Errorf("didn't expect ID to be 0: %w", gitprovider.ErrUnexpectedEvent)
+	}
+
+	return dk.c.c.DeleteToken(getRepoPath(dk.c.ref), dk.k.ID)
+}
+
+// Reconcile makes sure the desired state in this object (called "req" here) becomes
+// the actual state in the backing Git provider.
+//
+// If req doesn't exist under the hood, it is created (actionTaken == true).
+// If req doesn't equal the actual state, the resource will be updated (actionTaken == true).
+// If req is already the actual state, this is a no-op (actionTaken == false).
+//
+// The internal API object will be overridden with the received server data if actionTaken == true.
+func (dk *deployToken) Reconcile(ctx context.Context) (bool, error) {
+	// Always update for now.
+	// The problem here is that there is no way to retrieve the secret token value
+	// from GitLab again, so we'll never be sure if it actually matches what we have
+	// in this deployToken, so lets just update.
+	return true, dk.Update(ctx)
```

Comment:
We should move this up to the function doc to let users know what's actually going on.

Reply:
added 🎉